File: TheaterSuggestions/CSharp/SeatingPlaceSuggestions/SuggestionsAreMade.cs
namespace SeatsSuggestions;

/// <summary>
///     Occurs when a bunch of Suggestion are made.
/// </summary>
public class SuggestionsAreMade
{
    public SuggestionsAreMade(string showId, int partyRequested)
    {
        ShowId = showId;
        PartyRequested = partyRequested;

        InstantiateAnEmptyListForEveryPricingCategory();
    }

    public string ShowId { get; }
    public int PartyRequested { get; }

    private Dictionary<PricingCategory, List<SuggestionIsMade>> ForCategory { get; } = new();

    public IEnumerable<string> SeatNames(PricingCategory pricingCategory)
    {
        return ForCategory[pricingCategory].SelectMany(s => s.SeatNames());
    }

    private void InstantiateAnEmptyListForEveryPricingCategory()
    {
        foreach (PricingCategory pricingCategory in Enum.GetValues(typeof(PricingCategory)))
            ForCategory[pricingCategory] = new List<SuggestionIsMade>();
    }

    public void Add(IEnumerable<SuggestionIsMade> suggestions)
    {
        foreach (var suggestionMade in suggestions) ForCategory[suggestionMade.PricingCategory].Add(suggestionMade);
    }

    public bool MatchExpectations()
    {
        return ForCategory.SelectMany(s => s.Value).Any(x => x.MatchExpectation());
    }
}

File: TheaterSuggestions/Java/ScreeningSeatingArrangementSuggestions/src/main/java/org/weaveit/seatingplacesuggestions/SuggestionsAreMade.java
package org.weaveit.seatingplacesuggestions;

import java.util.*;

public class SuggestionsAreMade {
    private final String showId;
    private final int partyRequested;

    private final Map<PricingCategory, List<SuggestionIsMade>> forCategory = new HashMap<>();

    public SuggestionsAreMade(String showId, int partyRequested) {
        this.showId = showId;
        this.partyRequested = partyRequested;

        instantiateAnEmptyListForEveryPricingCategory();
    }

    public List<String> seatNames(PricingCategory pricingCategory) {
        return forCategory.get(pricingCategory).stream().map(SuggestionIsMade::seatNames).flatMap(Collection::stream).toList();
    }

    private void instantiateAnEmptyListForEveryPricingCategory() {
        for (PricingCategory pricingCategory : PricingCategory.values()) {
            forCategory.put(pricingCategory, new ArrayList<>());
        }
    }

    public void add(Iterable<SuggestionIsMade> suggestions) {
        suggestions.forEach(suggestionIsMade -> forCategory.get(suggestionIsMade.pricingCategory()).add(suggestionIsMade));
    }

    public boolean matchExpectations() {
        return forCategory.values().stream().flatMap(List::stream).anyMatch(SuggestionIsMade::matchExpectation);
    }

    public String showId() {
        return showId;
    }

    public int partyRequested() {
        return partyRequested;
    }
}
